code/ai/aicode.cpp

```diff
@@ -6159,6 +6168,81 @@ int check_ok_to_fire(int objnum, int target_objnum, weapon_info *wip)
 	return 1;
 }
 
+//	--------------------------------------------------------------------------
+//  Returns true if *aip has a line of sight to its current target.
+//	threshold defines the minimum radius for an object to be considered relevant for LoS
+bool check_los(int objnum, int target_objnum, float threshold) {
+	vec3d& end = Objects[target_objnum].pos;
+	vec3d& start = Objects[objnum].pos;
+
+	object* objp;
+
+	for (objp = GET_FIRST(&obj_used_list); objp != END_OF_LIST(&obj_used_list); objp = GET_NEXT(objp)) {
+		//Don't collision check against ourselves or our target
+		if (OBJ_INDEX(objp) == objnum || OBJ_INDEX(objp) == target_objnum)
+			continue;
+
+		int model_num = 0;
+		int model_instance_num = 0;
+
+		//Only collision check against other pieces of Debris, Asteroids or Ships
+		char type = objp->type;
+		if (type == OBJ_DEBRIS) {
+			model_num = Debris[objp->instance].model_num;
+			model_instance_num = -1;
+		}
+		else if (type == OBJ_ASTEROID) {
+			model_num = Asteroid_info[Asteroids[objp->instance].asteroid_type].model_num[Asteroids[objp->instance].asteroid_subtype];
+			model_instance_num = Asteroids[objp->instance].model_instance_num;
+		}
+		else if (type == OBJ_SHIP) {
+			model_num = Ship_info[Ships[objp->instance].ship_info_index].model_num;
+			model_instance_num = Ships[objp->instance].model_instance_num;
+		}
+		else
+			continue;
+
+		//Early Out Model too small
+
+		float radius = model_get(model_num)->rad;
+		if (radius < threshold)
+			continue;
+
+		vec3d a, b, c;
+		vm_vec_sub(&a, &start, &objp->pos);
+		vm_vec_sub(&b, &end, &start);
+		vm_vec_cross(&c, &b, &a);
+
+		float distToTargetRecip = 1.0f / vm_vec_dist_squared(&start, &end);
+		float distLoSSquared = vm_vec_mag_squared(&c) * distToTargetRecip;
+		float t = -vm_vec_dot(&a, &b) * distToTargetRecip;
+
+		radius *= radius;
+		float maxTdelta = sqrtf(radius * distToTargetRecip);
+		
+		//Early out Model too far from LoS
+		if (distLoSSquared > radius || -maxTdelta > t || maxTdelta + 1 < t)
+			continue;
```

Comment:
I'll be honest, I'm not really able to follow what exactly you're doing here, this is more or less what I had in mind:

```suggestion
    // if objp is inside start or end then we have to check the model
		if (vm_vec_dist(&start, &objp->pos) > objp->radius && vm_vec_dist(&end, &objp->pos) > objp->radius){
		    // now check that objp is in between start and end
		    vec3d start2end = end - start;
		    vec3d end2start = start - end;
		    vec3d start2objp = objp->pos - start;
		    vec3d end2objp = objp->pos - end;
		    
		    // if objp and end are in opposite directions from start, then early out
		    if(vm_vec_dot(&start2end, &start2objp) < 0.0f)
		        continue;
		         
		    // if objp and start are in opposite directions from end, then early out
		    if(vm_vec_dot(&end2start, &end2objp) < 0.0f)
		        continue; 
		        
		    // finally check if objp is too close to the path
		    if (vm_vec_mag(&start2objp) >  (vm_vec_mag(&start2end) + objp->radius))
		        continue; // adding objp->radius is somewhat of an overestimate but thats ok
		}
```

I hope I'm not being too 'micro-managey'

Reply:
This was essentially an implementation of https://mathworld.wolfram.com/Point-LineDistance3-Dimensional.html
And it then checks the distance from the Line of Sight to be checked, and if it is between both points (or within it's radius from the edges), creating a probably quite fast "is point in cylinder"-check. 
But if you want, I can use your code as well, resulting in a more readable source.
As for which is faster / better, I don't really know.

Reply:
Ah ok, I think a simpler, more easily-unstandable-at-a-glance approach would be better. We don't need an exact solution, just something that can quickly weed out 95% of cases, speed for both is probably pretty similar

Reply:
The functions `vm_vec_dist_to_line` and `vm_vec_dist_squared_to_line` can be found in vecmat.cpp.  It is better to use existing functions unless there is a clear advantage in writing a new one.

Reply:
I originally didn't use the functions as I could reuse lots of the intermediates for the second check, but as I've replaced the code anyway with Asteroth's suggestion, I guess it's fine anyways.
But yeah, maybe I should've used the vecmat implementation